File: rust/src/bin/1422_max_score.rs
// https://leetcode.com/problems/maximum-score-after-splitting-a-string/
// 01/01/2025

#[derive(Debug)]
struct Solution {}

impl Solution {
    pub fn max_score(s: String) -> i32 {
        let bytes = s.as_bytes();
        bytes
            .iter()
            .take(bytes.len() - 1)
            .enumerate()
            .map(|(i, _)| {
                let (left, right) = bytes.split_at(i + 1);
                (left.iter().filter(|&&b| b == b'0').count()
                    + right.iter().filter(|&&b| b == b'1').count()) as i32
            })
            .max()
            .unwrap_or(0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case("011101", 5)]
    #[case("00111", 5)]
    #[case("1111", 3)]
    fn test_max_score(#[case] s: String, #[case] expected: i32) {
        assert_eq!(Solution::max_score(s), expected);
    }
}

#[cfg(not(test))]
fn main() {
    let test_cases = vec![("011101", 5), ("00111", 5), ("1111", 3)];

    for (input, expected) in test_cases {
        let result = Solution::max_score(input.to_string());
        println!("Input: {}, Expected: {}, Got: {}", input, expected, result);
        assert_eq!(result, expected);
    }
}

File: elixir/1422_max_score.exs
# https://leetcode.com/problems/maximum-score-after-splitting-a-string/
# 01/01/2025

defmodule Solution do
  @spec max_score(s :: String.t()) :: integer
  def max_score(s) do
    total_ones = s |> String.graphemes() |> Enum.count(&(&1 == "1"))

    {max_score, _} =
      s
      |> String.graphemes()
      |> Enum.with_index(1)
      |> Enum.take(String.length(s) - 1)
      |> Enum.reduce({0, 0}, fn {char, idx}, {max, zeros} ->
        current_zeros = zeros + if(char == "0", do: 1, else: 0)
        left_ones = idx - current_zeros
        right_ones = total_ones - left_ones
        score = current_zeros + right_ones
        {max(max, score), current_zeros}
      end)

    max_score
  end
end

defmodule Main do
  def main do
    test_cases = [
      %{
        input: "011101",
        expected: 5
      },
      %{
        input: "00111",
        expected: 5
      },
      %{
        input: "1111",
        expected: 3
      }
    ]

    Enum.each(test_cases, fn %{input: input, expected: expected} ->
      result = Solution.max_score(input)
      IO.puts("Input: #{inspect(input)}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Result: #{if result == expected, do: "PASSED", else: "FAILED"}")
      IO.puts("------")
    end)
  end
end

Main.main()
